varinfo/cf_config.py

```diff
@@ -160,40 +155,43 @@ def _create_attributes_object(cf_item: dict) -> dict[str, str]:
             for attribute in cf_item.get('Attributes', {})
         }
 
-    def get_cf_attributes(
-        self, variable: str | None = None
-    ) -> dict[str, dict[str, str]]:
-        """Return the CF overrides and supplements that match a given
-        variable. If a variable is not specified, then return all overrides
-        and supplements. If there are no overrides or supplements, then
-        empty dictionaries will be returned instead.
```

Comment:
2) Precedence of conflicting rules. This is a sneaky bug that I only spotted this time around. If you have a variable name of "/this/is/my/variable" and had the following overrides:

```
{
    '/this/is/my/variable': {
        'attribute_name': 'more_specific_match',
    },
    '/this/is/.*': {
        'attribute_name': 'more_generic_match',
    }
}
```
Because the code used to just iterate through the rules and use the `dict.update` method, the last matching rule for a given attribute would take precedence. The ordering of the configuration file rules is somewhat arbitrary, and in the example above, this means the more generic rule would win out.

What now happens is the applicability rules are ordered by the length of the `Variable_Pattern`, with the assumption that the longer an applicability pattern, the more likely it is to be specific. (This isn't foolproof with complicated regular expressions, but we don't have too many of those to date). It is assumed that the more specific a regular expression is, the more likely it should take precedence. E.g., if someone write an override for a single variable, that likely should be applied instead of an override for all variables in a group (otherwise, why have the first rule at all).

I don't think this is necessarily perfect, but it feels better and more predictable behaviour than before.

Reply:
Yes, I was wondering if this is a good enough solution.  The thing that made the most sense to me was basically reverse order of the file. So that the "deepest" match would be used.

This would still match the "more_generic_match":
``` json
{ 
  '/this/is/my/variable': { 'attribute_name': 'more_specific_match', }, 
  '/this/is/.*': { 'attribute_name': 'more_generic_match', } 
}
```

And this would match 'more_specific_match':
``` json
{ 
  '/this/is/.*': { 'attribute_name': 'more_generic_match', }, 
  '/this/is/my/variable': { 'attribute_name': 'more_specific_match', }
  
}
```


But I expect that people will have both not put a lot of thought into how they
have structured the file and it may get pretty large, so I think sorting by
levels is a decent solution.  I kinda talked myself out of having the file
order be the important decider and I've been chewing it a few days.

This is going to trip people up no matter what.

Reply:
> I think sorting by levels

I'm conflating two thoughts, in xarray they're sorting by depth, basically counting `/`s You're proxing specificity by length of regexp.  still seems ok.

Reply:
This is interesting. I was toying with some sort of depth element to the proxy. I guess there are two places I could see a pure length-based proxy falling down:

```
/variable_name

vs

/(variable|other_pattern).*
```

Or:

```
/group/variable
/(really|long|group|match).*
```

Doing something that first finds the patterns with the most slashes, and then finds the longest string of those patterns, would prevent the second instance, but we're still going to have problems with the first. But, hey, it would still be better. Will have a play.

Reply:
Okay - this might be a bit more confusing, but I think it's a little bit better:

[edf9db6](https://github.com/nasa/earthdata-varinfo/pull/36/commits/edf9db6e153c193546a7909d76876303c689f0f3)

TL;DR: Depth is now the primary measure of specificity, then string length breaks the tie within a given hierarchical depth.

Reply:
I think we are going to hope it does the right thing for most people and then point at the code when it doesn't, but this seems good to me.